File: 0020-valid-parentheses/solution.java
class Solution {
    public boolean isValid(String s) {
        
        //creating a stack
        Stack<Character> stk= new Stack();
        
        for(int i=0; i<s.length() ; i++){
            char ch=s.charAt(i);
                
                //if ch is opning brecket
            if(ch=='(' || ch=='{'  || ch=='['){
                 stk.push(ch);
            }
            
            //if ch is closing brackets
            else if(ch==')' || ch=='}' || ch==']'){
                 //if stack is full
                if(stk.isEmpty()){
                    return false;
                }
                else if(!stk.isEmpty()){
                    char top=stk.peek();
                    if( (ch==')'  && top=='(' )  ||
                         (ch=='}'  && top=='{')  ||
                         (ch==']'  && top=='[')  ){
                        stk.pop();
                    }
                    else{
                        return false;
                    }
                }    

            }
        }
        return stk.isEmpty();
                       
    }
}

File: 0020-valid-parentheses/solution.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> stk;

        for(char ch : s){
            if(ch == '(' || ch == '[' || ch == '{'){
                stk.push(ch);
            }else{
                if(stk.size() == 0) return false;
                char top = stk.top();

                 if((top == '(' && ch == ')')  ||   (top == '{' && ch == '}')  || 
                   (top == '[' && ch == ']') 
                ) stk.pop();
                else return false;
            }
        }

        return stk.size() == 0;
    }
};
